d2/src/test/java/com/linkedin/d2/balancer/util/canary/TestBasicCanaryDistributionProviderImpl.java

```diff
@@ -0,0 +1,210 @@
+            .setTargetApplicationsStrategyProperties(appsProperties);
+
+    CanaryDistributionProviderImplFixture fixture = new CanaryDistributionProviderImplFixture();
+
+    // NOT in target list
+    Assert.assertEquals(fixture.getSpiedImpl("appC", null).distribute(targetAppsStrategy),
+        CanaryDistributionProvider.Distribution.STABLE, "App not in target list should return stable.");
+
+    // in scope and target list
+    Assert.assertEquals(fixture.getSpiedImpl("appA", null, 38).distribute(targetAppsStrategy),
+        CanaryDistributionProvider.Distribution.CANARY,
+        "App in target list and hash result in canary scope should return canary.");
+
+    // not in scope and in target list
+    Assert.assertEquals(fixture.getSpiedImpl("appA", null, 50).distribute(targetAppsStrategy),
+        CanaryDistributionProvider.Distribution.STABLE,
+        "App in target list but hash result not in canary scope should return stable.");
+  }
+
+  /**
+   * Provide objects for testing edge cases
+   * @return a list of objects with structure: {
+   *  D2 canary distribution strategy
+   * }
+   */
+  @DataProvider(name = "getEdgeCaseStrategies")
+  public Object[][] getEdgeCaseStrategies()
+  {
+    return new Object[][]{
+        // percentage strategy missing properties
+        {new D2CanaryDistributionStrategy().setStrategy(StrategyType.PERCENTAGE)},
+        // target hosts strategy missing properties
+        {new D2CanaryDistributionStrategy().setStrategy(StrategyType.TARGET_HOSTS)},
+        // target apps strategy missing properties
+        {new D2CanaryDistributionStrategy().setStrategy(StrategyType.TARGET_APPLICATIONS)},
+        // strategy with disabled type
+        {new D2CanaryDistributionStrategy().setStrategy(StrategyType.DISABLED)},
+        // strategy with unknown type
+        {new D2CanaryDistributionStrategy().setStrategy(StrategyType.$UNKNOWN)}};
+  }
+
+  @Test(dataProvider = "getEdgeCaseStrategies")
+  public void testEdgeCases(D2CanaryDistributionStrategy strategy)
+  {
+    Assert.assertEquals(new CanaryDistributionProviderImplFixture().getSpiedImpl().distribute(strategy),
+        CanaryDistributionProvider.Distribution.STABLE, "Invalid strategies should return stable");
+  }
+
+  @Test
+  public void testHashDistribution()
+  {
+    // verify hash results are almost uniformly distributed
+    int bucketCount = 100;
+    int attemptCount = 1000;
+    int[] hits = new int[bucketCount]; // to count hits in each bucket 0 ~ 99.
+    String serviceName = "test-service";
+    String hostName = "lor1-app";
+    int bucketIdx;
+    int hashResult;
+    for (int i = 0; i < attemptCount; i++) {
+      hashResult = (serviceName + hostName + i).hashCode();
+      bucketIdx = abs(hashResult) % 100; // hash "test-service"+"lor1-appXXX"
+      hits[bucketIdx]++;
+    }
+
+    // compare each bucket hits with the average hit, verify offsets are < 4
```

Comment:
Using 4 seems like a large range in distribution (almost 50% right?). Does if fail if you decrease that? Ideally I'd think we'd want it to be within 10% which is the limit of the test as it's set up.

Reply:
The total skewed hits are actually < 7% (about 62 / 1000). Please see my updated test.

Reply:
Thanks, left some comments on the new test code - I'm not sure I follow/agree with the skew calculation.